sharpd/sharp_zebra.c

```diff
@@ -211,66 +211,217 @@ int sharp_install_lsps_helper(bool install_p, bool update_p,
+ */
+static bool route_add(const struct prefix *p, vrf_id_t vrf_id, uint8_t instance,
+		      uint32_t nhgid, const struct nexthop_group *nhg,
+		      const struct nexthop_group *backup_nhg)
+{
+	struct zapi_route api;
+	struct zapi_nexthop *api_nh;
+	struct nexthop *nh;
+	int i = 0;
+
+	memset(&api, 0, sizeof(api));
+	api.vrf_id = vrf_id;
+	api.type = ZEBRA_ROUTE_SHARP;
+	api.instance = instance;
+	api.safi = SAFI_UNICAST;
+	memcpy(&api.prefix, p, sizeof(*p));
+
+	SET_FLAG(api.flags, ZEBRA_FLAG_ALLOW_RECURSION);
+	SET_FLAG(api.message, ZAPI_MESSAGE_NEXTHOP);
+
+	/* Only send via ID if nhgroup has been successfully installed */
+	if (nhgid && sharp_nhgroup_id_is_installed(nhgid)) {
+		SET_FLAG(api.message, ZAPI_MESSAGE_NHG);
+		api.nhgid = nhgid;
+	} else {
+		for (ALL_NEXTHOPS_PTR(nhg, nh)) {
+			api_nh = &api.nexthops[i];
+
+			zapi_nexthop_from_nexthop(api_nh, nh);
+
+			i++;
+		}
+		api.nexthop_num = i;
+	}
+
+	/* Include backup nexthops, if present */
+	if (backup_nhg && backup_nhg->nexthop) {
+		SET_FLAG(api.message, ZAPI_MESSAGE_BACKUP_NEXTHOPS);
+
+		i = 0;
+		for (ALL_NEXTHOPS_PTR(backup_nhg, nh)) {
+			api_nh = &api.backup_nexthops[i];
+
+			zapi_backup_nexthop_from_nexthop(api_nh, nh);
+
+			i++;
+		}
+
+		api.backup_nexthop_num = i;
+	}
+
+	if (zclient_route_send(ZEBRA_ROUTE_ADD, zclient, &api)
+	    == ZCLIENT_SEND_BUFFERED)
+		return true;
+	else
+		return false;
+}
+
+/*
+ * route_delete - Encodes a route for deletion to zebra
+ *
+ * This function returns true when the route sent was
+ * buffered by the underlying stream system.
+ */
+static bool route_delete(struct prefix *p, vrf_id_t vrf_id, uint8_t instance)
+{
+	struct zapi_route api;
+
+	memset(&api, 0, sizeof(api));
+	api.vrf_id = vrf_id;
+	api.type = ZEBRA_ROUTE_SHARP;
+	api.safi = SAFI_UNICAST;
+	api.instance = instance;
+	memcpy(&api.prefix, p, sizeof(*p));
+
+	if (zclient_route_send(ZEBRA_ROUTE_DELETE, zclient, &api)
+	    == ZCLIENT_SEND_BUFFERED)
+		return true;
+	else
+		return false;
+}
+
+static void sharp_install_routes_restart(struct prefix *p, uint32_t count,
+					 vrf_id_t vrf_id, uint8_t instance,
+					 uint32_t nhgid,
+					 const struct nexthop_group *nhg,
+					 const struct nexthop_group *backup_nhg,
+					 uint32_t routes)
 {
 	uint32_t temp, i;
 	bool v4 = false;
 
-	zlog_debug("Inserting %u routes", routes);
-
 	if (p->family == AF_INET) {
 		v4 = true;
 		temp = ntohl(p->u.prefix4.s_addr);
 	} else
 		temp = ntohl(p->u.val32[3]);
 
-	/* Only use backup route/nexthops if present */
-	if (backup_nhg && (backup_nhg->nexthop == NULL))
-		backup_nhg = NULL;
-
-	monotime(&sg.r.t_start);
-	for (i = 0; i < routes; i++) {
-		route_add(p, vrf_id, (uint8_t)instance, nhgid, nhg, backup_nhg);
+	for (i = count; i < routes; i++) {
+		bool buffered = route_add(p, vrf_id, (uint8_t)instance, nhgid,
+					  nhg, backup_nhg);
 		if (v4)
 			p->u.prefix4.s_addr = htonl(++temp);
 		else
 			p->u.val32[3] = htonl(++temp);
+
+		if (buffered) {
+			wb.p = *p;
```

Comment:
since `wb` is global, I'm just a bit concerned that it's fragile that no one ever clears/memsets it. I can't come up with path that fails because of that, where stale attributes are misleading after a callback - can you think of a path where that could happen?

Reply:
no not really.  Not concerned too much it's a testing daemon after all.